sql/src/main/java/io/crate/planner/node/management/ExplainPlan.java

```diff
@@ -85,12 +127,82 @@ public void execute(DependencyCarrier executor,
+                                                               NodeOperationTree operationTree) {
+        context.stopAndAddTimer(timerToken);
+        return (ignored, t) -> collectAndFinish(executor, consumer, jobId, operationTree, t);
+    }
+
+    private void collectAndFinish(DependencyCarrier executor,
+                                  RowConsumer consumer,
+                                  UUID jobId,
+                                  NodeOperationTree operationTree,
+                                  @Nullable Throwable executionFailure) {
+        if (executionFailure == null) {
+            collectTimingResults(jobId, executor, operationTree.nodeOperations())
+                .whenComplete(((nodeTimings, collectFailure) -> {
+                    if (collectFailure == null) {
+                        Map<String, Object> timings = buildResponseMap(context.getAsMap(), nodeTimings);
+                        consumer.accept(InMemoryBatchIterator.of(new Row1(timings), SENTINEL), null);
+                    } else {
+                        consumer.accept(null, collectFailure);
+                    }
+                }));
+        } else {
+            consumer.accept(null, executionFailure);
+        }
+    }
+
+    private TransportCollectProfileOperation getTransportCollectProfileOperation(DependencyCarrier executor, UUID jobId) {
+        TransportCollectProfileNodeAction nodeAction = executor.transportActionProvider()
+            .transportCollectProfileNodeAction();
+        return new TransportCollectProfileOperation(nodeAction, jobId);
+    }
+
+    private Map<String, Object> buildResponseMap(Map<String, Long> apeTimings, Map<String, Object> nodeTimings) {
+        MapBuilder<String, Object> mapBuilder = MapBuilder.newMapBuilder();
+        apeTimings.forEach(mapBuilder::put);
+        mapBuilder.put(Phase.Execute.name(), MapBuilder.<String, Object>newMapBuilder()
+            .put("Total", apeTimings.get(Phase.Execute.name()))
+            .putAll(nodeTimings)
+            .immutableMap());
+        return mapBuilder.immutableMap();
+    }
+
+    private CompletableFuture<Map<String, Object>> collectTimingResults(UUID jobId,
+                                                                        DependencyCarrier executor,
+                                                                        Collection<NodeOperation> nodeOperations) {
+        CompletableFuture<Map<String, Object>> resultFuture = new CompletableFuture<>();
+        Set<String> nodeIds = NodeOperationGrouper.groupByServer(nodeOperations).keySet();
+        TransportCollectProfileOperation collectProfileOperation = getTransportCollectProfileOperation(executor, jobId);
+
+        ConcurrentHashMap<String, Object> mergedMap = new ConcurrentHashMap<>(nodeIds.size());
+        AtomicInteger counter = new AtomicInteger(nodeIds.size());
+
+        for (String nodeId : nodeIds) {
+            collectProfileOperation.collect(nodeId)
```

Comment:
Maybe it would make sense to use `MultiBiConsumer` here. Although the current implementation is probably more efficient. Not sure what's clearer